### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load
purchase_csv = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(purchase_csv)
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Display the total number of players


In [2]:
# calculate total number of unique players
total_players = purchase_data["SN"].unique()
print(f"Total Number of Players = {len(total_players)}")

Total Number of Players = 576


## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [3]:
# calculate number of unique items purchased
unique_items = purchase_data["Item Name"].unique()
total_unique_items = len(unique_items)

# calculate average purchase price
avg_price = purchase_data["Price"].mean()

# calculate total number of purchases
total_purchase = purchase_data["Purchase ID"].count()

# calculate total revenue
total_revenue = purchase_data["Price"].sum()

In [4]:
# create summary data frame
summary_df = pd.DataFrame({"Number of Unique Items":[total_unique_items],
                          "Average Price":[avg_price],
                          "Number of Purchases":total_purchase,
                          "Total Revenue":total_revenue
                          })

# format values in summary data frame
summary_df["Average Price"] = summary_df["Average Price"].map("${:.2f}".format)
summary_df["Total Revenue"] = summary_df["Total Revenue"].map("${:,.2f}".format)

# display formatted summary data frame
summary_df

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [5]:
# drop duplicate player names to get a new data frame of unique players
cleaned_data = purchase_data.drop_duplicates(subset=["SN"])
cleaned_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [6]:
# get unique value counts for gender column and storing it in a new series
gender_count = cleaned_data["Gender"].value_counts()

In [7]:
# calculate count and percentage of male players
male_players = gender_count.loc["Male"]
percent_male_players = male_players/len(total_players)*100

# calculate count and percentage of female players
female_players = gender_count.loc["Female"]
percent_female_players = female_players/len(total_players)*100

# calculate count and percentage of other or non-disclosed players
other_players = gender_count.loc["Other / Non-Disclosed"]
percent_other_players = other_players/len(total_players)*100

In [8]:
# create and format gender summary table based on the calculations and set index
gender_summary = pd.DataFrame({"Total Count": [male_players, female_players, other_players], 
                               "Percentage of Players": [percent_male_players, percent_female_players, percent_other_players]}, 
                                index = ["Male", "Female", "Other / Non-Disclosed"])

# format values in summary data frame
gender_summary["Percentage of Players"] = gender_summary["Percentage of Players"].map("{:.2f}%".format)

# display formatted summary data frame
gender_summary

Total Count Percentage of Players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [9]:
# grouping purchase_data dataframe by Gender
purchase_by_gender = purchase_data.groupby("Gender")

# calculate purchase count
purchase_count = purchase_by_gender["Purchase ID"].count()

# calculate average purchase price
avg_purchase_price = purchase_by_gender["Price"].mean()

# calculate total purchase
total_purchase = purchase_by_gender["Price"].sum()

# calculate average purchase per person
avg_total_purchase_per_person = total_purchase/gender_count

In [10]:
# create purchase analysis summary table by gender
purchase_analysis_gender = pd.DataFrame({"Purchase Count": purchase_count,
                                         "Average Purchase Price":avg_purchase_price,
                                         "Total Purchase Value": total_purchase,
                                         "Avg Total Purchase per Person":avg_total_purchase_per_person})

# rename index as Gender
purchase_analysis_gender = purchase_analysis_gender.rename_axis('Gender')

# format values in purchase analysis summary table
purchase_analysis_gender["Average Purchase Price"] = purchase_analysis_gender["Average Purchase Price"].map("${:.2f}".format)
purchase_analysis_gender["Total Purchase Value"] = purchase_analysis_gender["Total Purchase Value"].map("${:,.2f}".format)
purchase_analysis_gender["Avg Total Purchase per Person"] = purchase_analysis_gender["Avg Total Purchase per Person"].map("${:.2f}".format)

#display purchase analysis summary table
purchase_analysis_gender

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                             $1,967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [11]:
# define bins and labels
age_bins = [0,9,14,19,24,29,34,39,50]
age_labels = ["<10","10-14","15-19","20-24","25-29","30-34","35-39","40+"]

In [12]:
# perform binning on age column in cleaned_data dataframe and add labels to existing data frame as a new column
cleaned_data["Age Group"] = pd.cut(cleaned_data.iloc[:,2], age_bins, labels = age_labels)
cleaned_data.head()

<ipython-input-12-dc7ae4e2854b>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cleaned_data["Age Group"] = pd.cut(cleaned_data.iloc[:,2], age_bins, labels = age_labels)


Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price Age Group  
0  Extraction, Quickblade Of Trembling Hands   3.53     20-24  
1                          Frenzied Scimitar   1.56       40+  
2                               Final Critic   4.88     20-24  
3                                Blindscythe   3.27     20-24  
4                                       Fury   1.44     20-24

In [13]:
# groupby labels and set them as index
age_index = cleaned_data.groupby("Age Group")

# calculate total count age in each bin
total_count_age = age_index["Purchase ID"].count()

# calculate percentage of players in each bin
percent_players = total_count_age/len(total_players)*100

In [14]:
# create age demographics table
age_demographics = pd.merge(total_count_age, percent_players, on="Age Group")

# rename columns in the dataframe
age_demographics = age_demographics.rename(columns={
    "Purchase ID_x":"Total Count", 
    "Purchase ID_y":"Percentage of Players"})


# format value in age demographics table
age_demographics["Percentage of Players"] = age_demographics["Percentage of Players"].map("{:.2f}%".format)

# remove index name
age_demographics.index.name = None

#display age demographics table
age_demographics

Total Count Percentage of Players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [15]:
# perform binning on age column in purchase_data dataframe and add labels to existing data frame as a new column
purchase_data["Labels"] = pd.cut(purchase_data["Age"], age_bins, labels = age_labels)
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price Labels  
0  Extraction, Quickblade Of Trembling Hands   3.53  20-24  
1                          Frenzied Scimitar   1.56    40+  
2                               Final Critic   4.88  20-24  
3                                Blindscythe   3.27  20-24  
4                                       Fury   1.44  20-24

In [24]:
# groupby labels and set them as index
purchase_data_by_age = purchase_data.groupby("Labels")

In [25]:
# calculate purchase count
purchase_count_age = purchase_data_by_age["Purchase ID"].count()

# calculate average purchase price
avg_purchase_price_age = purchase_data_by_age["Price"].mean()

# calculate total purchase
total_purchase_age = purchase_data_by_age["Price"].sum()

# calculate average purchase per person by dividing total purchase in each bin by unique total count in each bin
avg_total_purchase_per_person_age = total_purchase_age/total_count_age

In [26]:
# merge purchase count age and average purchase price dataframe into a new dataframe
purchase_analysis_by_age = pd.merge(purchase_count_age, avg_purchase_price_age, on="Labels")

# merge total purchase dataframe into purchase analysis by age dataframe
purchase_analysis_by_age = pd.merge(purchase_analysis_by_age, total_purchase_age, on="Labels")

# add average total purchase per person age series as the column in dataframe
purchase_analysis_by_age ["Avg Total Purchase per Person"] = avg_total_purchase_per_person_age

# rename columns
purchase_analysis_by_age = purchase_analysis_by_age.rename(columns={
    "Purchase ID":"Purchase Count", 
    "Price_x":"Average Purchase Price", 
    "Price_y":"Total Purchase Value"})

# rename axis
purchase_analysis_by_age = purchase_analysis_by_age.rename_axis('Age Ranges')

# format values in purchase analysis summary table
purchase_analysis_by_age["Average Purchase Price"] = purchase_analysis_by_age["Average Purchase Price"].map("${:.2f}".format)
purchase_analysis_by_age["Total Purchase Value"] = purchase_analysis_by_age["Total Purchase Value"].map("${:,.2f}".format)
purchase_analysis_by_age["Avg Total Purchase per Person"] = purchase_analysis_by_age["Avg Total Purchase per Person"].map("${:.2f}".format)

# display purchase analysis by age summary table
purchase_analysis_by_age

Purchase Count Average Purchase Price Total Purchase Value  \
Age Ranges                                                               
<10                     23                  $3.35               $77.13   
10-14                   28                  $2.96               $82.78   
15-19                  136                  $3.04              $412.89   
20-24                  365                  $3.05            $1,114.06   
25-29                  101                  $2.90              $293.00   
30-34                   73                  $2.93              $214.00   
35-39                   41                  $3.60              $147.67   
40+                     13                  $2.94               $38.24   

           Avg Total Purchase per Person  
Age Ranges                                
<10                                $4.54  
10-14                              $3.76  
15-19                              $3.86  
20-24                              $4.32  
25-29                              $3.81  
30-34                              $4.12  
35-39                              $4.76  
40+                                $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [30]:
# groupby purchase data by unique player names
purchase_by_players = purchase_data.groupby("SN")

In [31]:
# calculate purchase count
purchase_count_players = purchase_by_players["Purchase ID"].count()

# calculate average purchase price
avg_price_players = purchase_by_players["Price"].mean()

# calculate total purchase price
total_purchase_players = purchase_by_players["Price"].sum()

In [33]:
# merge purchase count and average price for each player into a new data frame
player_combined_1 = pd.merge(purchase_count_players, avg_price_players, on="SN")

# merge total purchase for each player with player combined 1 data frame into player spending dataframe
player_spending = pd.merge(player_combined_1, total_purchase_players, on="SN")

# display data frame which has purchase count, average price and total purchase price for each player
player_spending.head()

Purchase ID  Price_x  Price_y
SN                                         
Adairialis76            1     2.28     2.28
Adastirin33             1     4.48     4.48
Aeda94                  1     4.91     4.91
Aela59                  1     4.32     4.32
Aelaria33               1     1.79     1.79

In [36]:
# perform a descending sort on column Price_y which is total purchase file to find the top 5 spender
top_spender = player_spending.sort_values("Price_y", ascending=False)

In [37]:
# rename columns in top spender data frame
top_spender = top_spender.rename(columns={"Purchase ID":"Purchase Count",
                                          "Price_x":"Average Purchase Price",
                                          "Price_y":"Total Purchase Value"})

# format values in top spender data frame
top_spender["Average Purchase Price"] = top_spender["Average Purchase Price"].map("${:.2f}".format)
top_spender["Total Purchase Value"] = top_spender["Total Purchase Value"].map("${:.2f}".format)

# display top 5 spender
top_spender.head()

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, average item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [38]:
# retreive only Item ID, Item Name and Price columns in a new data frame
purchase_data_item = purchase_data.iloc[:,4:]
purchase_data_item.head()

Item ID                                  Item Name  Price Labels
0      108  Extraction, Quickblade Of Trembling Hands   3.53  20-24
1      143                          Frenzied Scimitar   1.56    40+
2       92                               Final Critic   4.88  20-24
3      100                                Blindscythe   3.27  20-24
4      131                                       Fury   1.44  20-24

In [40]:
# groupby using Item ID and Item Name
item_id_name = purchase_data_item.groupby(["Item ID","Item Name"])

In [41]:
# calculate purchase count for each item name and item id
item_purchase_count = item_id_name["Price"].count()

# calculate item price for each item name and item id
item_price = item_id_name["Price"].mean()

# calculate total purchase for each item name and item id
item_total_purchase = item_id_name["Price"].sum()

In [42]:
# merge purchase count and item price values in a new data frame item combined1
item_combined1 = pd.merge(item_purchase_count, item_price, on=("Item ID","Item Name"))

# rename columns
item_combined1 = item_combined1.rename(columns = {"Price_x":"Purchase Count",
                                                  "Price_y":"Item Price"})

item_combined1.head()

Purchase Count  Item Price
Item ID Item Name                                     
0       Splinter                         4      1.2800
1       Crucifer                         4      2.9425
2       Verdict                          6      2.4800
3       Phantomlight                     6      2.4900
4       Bloodlord's Fetish               5      1.7000

In [43]:
# merge previous data frame and total purchase into a item combined final dataframe
item_combined_final = pd.merge(item_combined1, item_total_purchase, on=("Item ID","Item Name"))

# rename column
item_combined_final = item_combined_final.rename(columns = {"Price":"Total Purchase Value"})

# display combined dataframe with all the three columns
item_combined_final.head()

Purchase Count  Item Price  Total Purchase Value
Item ID Item Name                                                           
0       Splinter                         4      1.2800                  5.12
1       Crucifer                         4      2.9425                 11.77
2       Verdict                          6      2.4800                 14.88
3       Phantomlight                     6      2.4900                 14.94
4       Bloodlord's Fetish               5      1.7000                  8.50

In [44]:
# copy combined dataframe into a new dataframe for formatting
item_combined_format = item_combined_final.copy()

# format values in the data frame
item_combined_format["Item Price"] = item_combined_format["Item Price"].map("${:.2f}".format)
item_combined_format["Total Purchase Value"] = item_combined_format["Total Purchase Value"].map("${:.2f}".format)

# display formatted data frame
item_combined_format.head()

Purchase Count Item Price Total Purchase Value
Item ID Item Name                                                         
0       Splinter                         4      $1.28                $5.12
1       Crucifer                         4      $2.94               $11.77
2       Verdict                          6      $2.48               $14.88
3       Phantomlight                     6      $2.49               $14.94
4       Bloodlord's Fetish               5      $1.70                $8.50

In [45]:
# perform descending sort on column Purchase Count in formatted dataframe to find the top 5 popular items purchased by the players
popular_item = item_combined_format.sort_values("Purchase Count", ascending=False)

popular_item.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [46]:
# perform descending sort on column Total Purchase Value in Item Combined Final dataframe to find most profitable items purchased by the players
profitable_item = item_combined_final.sort_values("Total Purchase Value", ascending=False)

# format values in the data frame
profitable_item["Item Price"] = profitable_item["Item Price"].map("${:.2f}".format)
profitable_item["Total Purchase Value"] = profitable_item["Total Purchase Value"].map("${:.2f}".format)


profitable_item.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

## Three Observable Trends from the Heroes of Pymoli data analysis are:

1. From Age Demographics analyis it can be understood that 44.79% of players playing this game belong to the age group of 20-24 followed by 18.58% in age group of 15-19. This means that at least 63.37% of players are from younger generation.


2. 84.03% of players on this game are males who contribute to USD 1967.64 in revenue which is 82.68% of total revenue


3. Largest active age group of 20-24 on the game also contributes towards the most in purchasing optional items. They purchased optional items worth of USD 1114.06 which is 46.81% of total purchase value.


Based on these 3 observations, I think the gaming company should make an effort to target males in the age group of 20-24 with more new exciting optional items and encourage them to purchase more as this will certainly generate more revenue and profits for the company.